File: Python/MaxDepthBST.py
# Return the maximum depth of a binary search tree.

class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def insert(root, val):
    if root is None:
        return TreeNode(val)
    elif val < root.val:
        root.left = insert(root.left, val)
    else:
        root.right = insert(root.right, val)

    return root

def findDepth(head):
    if head is None:
        return 0
    return 1 + max(findDepth(head.left), findDepth(head.right))

def main():
    root = None

    root = insert(root, 5)
    insert(root, 3)
    insert(root, 8)
    insert(root, 2)
    insert(root, 4)
    insert(root, 7)

    print("BST created with root value:", root.val)
    print("Max Depth: ", findDepth(root))

File: Java/MaxDepthBST.java
// Return the maximum depth of a binary search tree.
class TreeNode{
    int val;
    TreeNode left, right;
    TreeNode(int val){
        this.val = val;
        left = right = null;
    }
}

class MaxDepthBST{
    TreeNode insert(TreeNode root, int val) {
        if(root == null) {return new TreeNode(val);}
        else if(val < root.val) {root.left = insert(root.left, val);}
        else {root.right = insert(root.right, val);}
        return root;
    }

    int findDepth(TreeNode head){
        if(head == null) return 0;
        return 1 + Math.max(findDepth(head.left), findDepth(head.right));
    }

    public static void main(String[] args) {
        MaxDepthBST bst = new MaxDepthBST();
        TreeNode root = null;
        root = bst.insert(root, 5);
        bst.insert(root, 3);
        bst.insert(root, 8);
        bst.insert(root, 2);
        bst.insert(root, 4);
        bst.insert(root, 7);
        System.out.println("BST created with root value: " + root.val);
        System.out.println("Max Depth: " + bst.findDepth(root));
    }
}


    

